### ISSUE TITLE
[java] 7.0.0-rc2 missing ASTClassOrInterfaceDeclaration visits

### ISSUE THREAD

**Issue by NilsBaumgartner1994** [pmd/pmd#4566]
**Affects PMD Version:** 

7.0.0-rc2 and 7.0.0-rc1

**Description:**
The amount of visited ASTClassOrInterfaceDeclaration changed from 6.55.0 to new version 7.0.0-rc2, altough the amount of visited files remains same.

PMD7: 1503 files
	ClassOrInterfaceDeclarations: 1589	
PMD6: 1503 files
	ClassOrInterfaceDeclarations: 1909

**Steps to reproduce:**
Analysed: ArgoUML v 0.26 Beta

```java
package net.sourceforge.pmd.examples.java.rules;

import ...

public class MyRule extends AbstractJavaRule {

    static int count = 0;

    private static final StringProperty BAD_NAME = StringProperty.named("badName")
            .defaultValue("foo")
            .desc("The variable name that should not be used.")
            .uiOrder(1.0f)
            .build();

    public MyRule() {
        definePropertyDescriptor(BAD_NAME);
    }

    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
        String className = node.getCanonicalName();
        System.out.println(count+" - "+className);
        ASTClassOrInterfaceType superClass = node.getSuperClassTypeNode();
        if(superClass!=null){
            System.out.println("extends: "+superClass.getSimpleName());
        }

        MyRule.count++;
        return null;
    }
}

```

Please provide detailed steps for how we can reproduce the bug.
```
.PHONY: all build run clean

all: clean build run

setup:
	./mvnw clean install

build:
	./mvnw clean install
	unzip -o ./pmd-java-dist/target/pmd-java-bin-1.0.0-SNAPSHOT.zip -d ./pmd-java-dist/target/
	cp ./pmd-java-dist/target/pmd-java-bin-1.0.0-SNAPSHOT/lib/pmd-java-custom-1.0.0-SNAPSHOT.jar ./pmd-bin-7.0.0-rc2/lib/

run:
	./pmd-bin-7.0.0-rc2/bin/pmd check -d /Users/nbaumgartner/Desktop/src -f text -R custom-java-ruleset.xml

clean:
	rm -rf ./pmd-java-dist/target/
	rm -rf ./pmd-java-custom/target/
```
**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

CLI

**Comment by jsotuyod** [comment]
@NilsBaumgartner1994 this is unexpected… we did plenty of changes to the AST, but I can't think of any that would affect this in particular… and we don't see a drop on reported issues either in our regression tester…

Since you rules prints the names of all matched files, can you share the outputs for both runs for an easier comparison?

Also, which beta of ArgoUML 0.26 did you analyze? I see there are 5 tagged betas on GH, the last being https://github.com/argouml-tigris-org/argouml/tree/VERSION_0_26_BETA_5/

**Comment by NilsBaumgartner1994** [comment]
> @NilsBaumgartner1994 this is unexpected… we did plenty of changes to the AST, but I can't think of any that would affect this in particular… and we don't see a drop on reported issues either in our regression tester…
> 
> 
> 
> Since you rules prints the names of all matched files, can you share the outputs for both runs for an easier comparison?
> 
> 
> 
> Also, which beta of ArgoUML 0.26 did you analyze? I see there are 5 tagged betas on GH, the last being https://github.com/argouml-tigris-org/argouml/tree/VERSION_0_26_BETA_5/

I used the one from the unified Bugdataset. I will further investigate and will share the project. Maybe i am missing compilation steps or something.

**Comment by adangel** [comment]
The difference might be inner classes. In your rule above, you are _not_ visiting any inner classes, since you "return null";. Did you maybe use rule chain in the PMD6 version of the rule?

**Comment by jsotuyod** [comment]
Yeah, but it also is not calling `super.visit(…)` which in PMD 6 had the same effect of not diving into the child nodes, and therefore omitting inner / local classes…

I guess the sane thing would be to try both with a properly written rule such as:


```java
public class MyRule extends AbstractJavaRule {

    // FIXME : Not thread safe, will only work when PMD is run with -t 0 or -t 1 (default)
    static int count = 0;

    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
        super.visit(node, data);
        String className = node.getCanonicalName();
        System.out.println(count+" - "+className);
        ASTClassOrInterfaceType superClass = node.getSuperClassTypeNode();
        if(superClass!=null){
            System.out.println("extends: "+superClass.getSimpleName());
        }

        MyRule.count++;
        return data;
    }
}
```

**Comment by NilsBaumgartner1994** [comment]
> The difference might be inner classes. In your rule above, you are _not_ visiting any inner classes, since you "return null";. Did you maybe use rule chain in the PMD6 version of the rule?

correct, that was the issue. I used the rule chaining in PMD6. Now when returning super.visit it finds the same classes.

**Comment by adangel** [comment]
Note that you can still use rule chain with pmd7: Inherit from `AbstractJavaRulechainRule` and provide `ASTClassOrInterfaceDeclaration.class` as the argument to the super constructor.
